### PR TITLE
Add UpsertTrustedClusterV2 RPC

### LINKED ISSUE #48309: `trusted_cluster` resource renames itself, breaking subsequent updates
### Expected behaviour:

- As a user I create a resource called `clusterA` while my remote cluster real name is `clusterA.example.com`
Either I have a name mismatch error when I create the resource, or the resource name stays `clusterA`

### Current behaviour:

- The resource is renamed from `clusterA` to `clusterA.example.com` in the backend
- I don't have a unique identifier to get the resource anymore
- If I edit the role mapping and apply again for `clusterA` it silently drop my changes

### Bug details:
- Teleport version v16.4.2
- Recreation steps: create a TC resource whose name is different than the root cluster name
- Debug logs

### Additional notes

- this is blocking TC support in the operator

### PR DESCRIPTION
**Pull request by bernardjkim**
Closes https://github.com/gravitational/teleport/issues/48309
Supports https://github.com/gravitational/teleport/issues/22474

This PR introduces the `UpsertTrustedClusterV2` rpc. This new rpc supersedes `UpsertTrustedCluster`. V2 performs resource name validation for trusted_clusters. This is required to support trusted_cluster resources with the kubernetes operator.

### Approaches Considered
Previously in https://github.com/gravitational/teleport/pull/49169, there was discussion about introducing a new TrustedClusterV3 to address https://github.com/gravitational/teleport/issues/48309. This approach could work, but it is unnecessary because V3 would not introduce any breaking changes in the resource itself. We would also need to be careful not to break backwards compatibility while introducing this new resource version.

We also considered using the resource origin or label to identify if resource name validation is required. This would allow the auth server to only validate the resource name if the trusted_cluster origin is kubernetes. However, this approach would not be able to ensure name validation if the kubernetes operator connects to an older auth server.

We've decided that a new endpoint was the more appropriate solution. This approach enables the auth server to ensure trusted_cluster resource names are validated when created by the kubernetes operator. The kubernetes operator would also not be able to bypass the resource name validation when connected to an older auth server.

### Deprecation/Migration
No current timeline to migrate other clients to use the `UpsertTrustedClusterV2`. `UpsertTrustedCluster` will be supported indefinitely for now.

### CODE REVIEW

**Comment by programmerq** [comment]
If a trusted cluster is upserted and labels are added/changed/modified, would that also propagate the label change to the corresponding `remote_cluster` object?

File: api/types/trustedcluster.go [hunk 1]
```diff
@@ -184,6 +184,16 @@ func (c *TrustedClusterV2) SetName(e string) {
...
+// Origin returns the origin value of the resource.
+func (c *TrustedClusterV2) Origin() string {
+	return c.Metadata.Origin()
+}
+
+// SetOrigin sets the origin value of the resource.
+func (c *TrustedClusterV2) SetOrigin(origin string) {
+	c.Metadata.SetOrigin(origin)
+}
+
```
[the remark this answers is not in the document]
Comment (bernardjkim) [reply api/types/trustedcluster.go:196]: I believe these are still needed to implement https://github.com/gravitational/teleport/blob/01fc9f0acfde26950c00076180c207aee77f194a/integrations/operator/controllers/reconcilers/legacy_resource_without_labels.go#L29-L37

**Comment by bernardjkim** [comment]
> If a trusted cluster is upserted and labels are added/changed/modified, would that also propagate the label change to the corresponding remote_cluster object?

According to the documentation, labels are not support on `trusted_cluster` resources. https://goteleport.com/docs/admin-guides/management/admin/trustedclusters/#step-56-manage-access-to-the-trusted-cluster

File: api/client/client.go [hunk 2]
```diff
@@ -2277,16 +2277,35 @@ func (c *Client) GetTrustedClusters(ctx context.Context) ([]types.TrustedCluster
...
+	if !ok {
+		return nil, trace.BadParameter("invalid type %T", trustedCluster)
 	}
-	return resp, nil
+	resp, err := c.grpc.UpsertTrustedClusterV2(ctx, trustedClusterV2)
+	if trace.IsNotImplemented(err) {
+		// Try to print a nicer error message when newer clients connect to
+		// older auth servers that don't recognize the new gRPC.
+		authVersion := "unknown"
+		if pingResp, err := c.Ping(ctx); err == nil && pingResp.ServerVersion != "" {
```
Comment (espadolini) [inline api/client/client.go:2302]: I don't know if this is worth it just to add some version number in the error message, compared to just writing "possible version mismatch between this client and the control plane".

If only the server version was made available in the unknown service handler... 😢

File: api/client/client.go [hunk 3]
```diff
@@ -2277,16 +2277,35 @@ func (c *Client) GetTrustedClusters(ctx context.Context) ([]types.TrustedCluster
...
+			authVersion = pingResp.ServerVersion
+		}
+		return resp, trace.Wrap(err, "client version (%s) is likely newer than your auth server version (%s), "+
```
Comment (espadolini) [inline api/client/client.go:2305]: "Auth Service" should be capitalized in user facing messages, but "control plane" is probably clearer.

Also, the client version can be higher, not newer - the client can be newer.

```suggestion
		return nil, trace.Wrap(err, "client (version %s) is likely newer than your control plane (version %s), "+
```

```suggestion
		return nil, trace.Wrap(err, "client version (%s) is likely higher than your control plane version (%s), "+
```

File: api/client/client.go [hunk 4]
```diff
@@ -2277,16 +2277,35 @@ func (c *Client) GetTrustedClusters(ctx context.Context) ([]types.TrustedCluster
...
+			"consider upgrading your auth server", api.Version, authVersion)
```
Comment (espadolini) [inline api/client/client.go:2306]: Do we really want to advise upgrading the auth like that? If this error message started appearing in some automation and SREs were scrambling for a fix, surely they'd rather downgrade the operator or tf provider rather than upgrade the cluster?

File: api/client/client.go [hunk 5]
```diff
@@ -2277,16 +2277,35 @@ func (c *Client) GetTrustedClusters(ctx context.Context) ([]types.TrustedCluster
...
+	trustedClusterV2, ok := trustedCluster.(*types.TrustedClusterV2)
 	if !ok {
-		return nil, trace.BadParameter("invalid type %T", trusedCluster)
+		return nil, trace.BadParameter("invalid type %T", trustedCluster)
 	}
-	resp, err := c.grpc.UpsertTrustedCluster(ctx, trustedCluster)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	resp, err := c.grpc.UpsertTrustedCluster(ctx, trustedClusterV2)
+	return resp, trace.Wrap(err)
```
Comment (espadolini) [inline api/client/client.go:2288]:
```suggestion
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return resp, nil
```

File: api/client/client.go [hunk 6]
```diff
@@ -2277,16 +2277,35 @@ func (c *Client) GetTrustedClusters(ctx context.Context) ([]types.TrustedCluster
...
+		// Try to print a nicer error message when newer clients connect to
+		// older auth servers that don't recognize the new gRPC.
+		authVersion := "unknown"
+		if pingResp, err := c.Ping(ctx); err == nil && pingResp.ServerVersion != "" {
+			authVersion = pingResp.ServerVersion
+		}
+		return resp, trace.Wrap(err, "client version (%s) is likely newer than your auth server version (%s), "+
+			"consider upgrading your auth server", api.Version, authVersion)
+	}
+	return resp, trace.Wrap(err)
```
Comment (espadolini) [inline api/client/client.go:2308]:
```suggestion
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return resp, nil
```

Comment (espadolini) [reply api/client/client.go:2308]: (the handling for NotImplemented could be put inside the `err != nil` case but that's not super important)

File: integration/integration_test.go [hunk 7]
```diff
@@ -2796,7 +2797,7 @@ func testMapRoles(t *testing.T, suite *integrationTestSuite) {
 	require.NoError(t, services.CheckAndSetDefaults(trustedCluster))
 
 	// try and upsert a trusted cluster
-	helpers.TryCreateTrustedCluster(t, aux.Process.GetAuthServer(), trustedCluster)
+	helpers.TryCreateTrustedCluster(t, aux.Process.GetAuthServer(), trustedCluster, false)
```
Comment (espadolini) [inline integration/integration_test.go:2800]:
```suggestion
	const skipNameValidationFalse = false
	helpers.TryCreateTrustedCluster(t, aux.Process.GetAuthServer(), trustedCluster, skipNameValidationFalse)
```

File: lib/auth/trustedcluster.go [hunk 8]
```diff
@@ -46,6 +46,19 @@ import (
 
 // UpsertTrustedCluster creates or toggles a Trusted Cluster relationship.
 func (a *Server) UpsertTrustedCluster(ctx context.Context, tc types.TrustedCluster) (newTrustedCluster types.TrustedCluster, returnErr error) {
+	upserted, err := a.upsertTrustedCluster(ctx, tc, false)
```
Comment (espadolini) [inline lib/auth/trustedcluster.go:49]:
```suggestion
	const validateNameFalse = false
	upserted, err := a.upsertTrustedCluster(ctx, tc, validateNameFalse)
```

File: lib/auth/trustedcluster.go [hunk 9]
```diff
@@ -46,6 +46,19 @@ import (
...
+	return upserted, trace.Wrap(err)
+}
+
+// UpsertTrustedClusterV2 creates or toggles a Trusted Cluster relationship.
+// The trusted cluster resource name must match the cluster name.
+func (a *Server) UpsertTrustedClusterV2(ctx context.Context, tc types.TrustedCluster) (newTrustedCluster types.TrustedCluster, returnErr error) {
+	upserted, err := a.upsertTrustedCluster(ctx, tc, true)
```
Comment (espadolini) [inline lib/auth/trustedcluster.go:56]:
```suggestion
	const validateNameTrue = true
	upserted, err := a.upsertTrustedCluster(ctx, tc, validateNameTrue)
```

File: lib/auth/trustedcluster.go [hunk 10]
```diff
@@ -322,9 +335,10 @@ func (a *Server) establishTrust(ctx context.Context, trustedCluster types.Truste
 			if remoteClusterName == domainName {
 				return nil, trace.BadParameter("remote cluster name can not be the same as local cluster name")
 			}
-			// TODO(klizhentas) in 2.5.0 prohibit adding trusted cluster resource name
-			// different from cluster name (we had no way of checking this before x509,
-			// because SSH CA was a public key, not a cert with metadata)
+			if validateName && trustedCluster.GetName() != remoteClusterName {
```
Comment (espadolini) [inline lib/auth/trustedcluster.go:338]: We should also check if the cluster we're connected to has the correct name _before_ establishing trust; bailing out here will still leave the remote_cluster in the root, and it would be nice to avoid that. Both proto.AuthServer/Ping on the auth and webapi/find on the proxy are usable by unauthenticated clients and contain the cluster name.

File: lib/auth/trustedcluster.go [hunk 11]
```diff
@@ -322,9 +335,10 @@ func (a *Server) establishTrust(ctx context.Context, trustedCluster types.Truste
...
+				return nil, trace.CompareFailed("trusted cluster resource name must be the same as the remote cluster name. got: %q, expected: %q",
```
Comment (espadolini) [inline lib/auth/trustedcluster.go:339]: Should we worry about revealing info that the caller should not know? The power to upsert a trusted cluster _is_ basically admin, anyway...

Comment (espadolini) [inline lib/auth/trustedcluster.go:339]: I'm not sure if CompareFailed is appropriate here, we tend to use that to signal that some operation should be restarted and tried again because of concurrency, but there's nothing automated that can fix this. Maybe BadParameter?

File: api/client/client.go (hunk 4 above)
Comment (bernardjkim) [reply api/client/client.go:2306]: If the operator receives a `NotImplemented` error, the only solution would be to update the Auth Service to a version that supports this new endpoint.

This shouldn't be a problem though, the teleport operator does not support trusted_clusters yet so it won't break any existing automation.

File: lib/auth/trustedcluster.go (hunk 10 above)
Comment (bernardjkim) [reply lib/auth/trustedcluster.go:354]: Good call, thanks for pointing this out.

File: api/types/trustedcluster.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (espadolini) [reply api/types/trustedcluster.go:196]: Does the origin mess with the checks we do to only allow updating the role map or the enabled/disabled state?

File: lib/auth/trustedcluster.go [hunk 12]
```diff
@@ -550,10 +580,54 @@ func (a *Server) validateTrustedClusterToken(ctx context.Context, tokenName stri
...
+		return nil, trace.Wrap(err)
+	}
+
+	return validateResponse, nil
+}
+
+// validateTrustedClusterName validates that the trusted cluster resource name
+// matches the cluster name.
+func (a *Server) validateTrustedClusterName(ctx context.Context, clt *roundtrip.Client, trustedCluster types.TrustedCluster) error {
+	out, err := httplib.ConvertResponse(clt.Get(ctx, clt.Endpoint("webapi", "ping"), nil))
```
Comment (espadolini) [inline lib/auth/trustedcluster.go:611]: Is there a way to reuse the `api/client/webclient.Find` code?

Also, `find` still has the cluster name so that's preferable to `ping`.

File: lib/auth/trustedcluster.go [hunk 13]
```diff
@@ -46,13 +47,15 @@ import (
...
+	const validateName = false
+	upserted, err := a.upsertTrustedCluster(ctx, tc, validateName)
 	return upserted, trace.Wrap(err)
 }
 
 // UpsertTrustedClusterV2 creates or toggles a Trusted Cluster relationship.
 // The trusted cluster resource name must match the cluster name.
 func (a *Server) UpsertTrustedClusterV2(ctx context.Context, tc types.TrustedCluster) (newTrustedCluster types.TrustedCluster, returnErr error) {
+	const validateName = true
 	upserted, err := a.upsertTrustedCluster(ctx, tc, true)
```
[the remark this answers is not in the document]
Comment (bernardjkim) [reply lib/auth/trustedcluster.go]: Oops. `validateNameTrue` is okay with me.

File: api/types/trustedcluster.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernardjkim) [reply api/types/trustedcluster.go:196]: Ahh, yeah. When the operator restarts it will attempt to propagate the origin label and the resource revision. The operator will then fail to reconcile the trusted_cluster with "leaf cluster is already enabled, this update would have no effect"
https://github.com/gravitational/teleport/blob/94f1291569478bf3cea0d5b8ba287657d6814bf8/api/types/trustedcluster.go#L237-L264

Do you know if there is a reason we do not allow an update if does not change either of those fields?

[the remark this answers is not in the document]
Comment (espadolini) [reply api/types/trustedcluster.go:196]: I think that we could/should allow updating labels, they don't affect anything else after all. Do IaC tools need a way to only overwrite the resource if its origin is the expected one, or are they fine with just upserting to keep the value up to date?

[the remark this answers is not in the document]
Comment (hugoShaka) [reply api/types/trustedcluster.go:196]: > Do IaC tools need a way to only overwrite the resource if its origin is the expected one

The operator will get the existing resource first and bail out if the resource doesn't have the right origin. Else it will update or upsert based on the resource. The upsert is racy, but this catches most issues.

Tl;dr we need the origin label support, we don't need conditional update based on the origin because we do it IaC side. However users should be allowed to update labels. For example you create a TC, then want the operator to manage it. For the operator to adopt the resource you must edit the resource and add the kubernetes origin.

[the remark this answers is not in the document]
Comment (bernardjkim) [reply api/types/trustedcluster.go:196]: That makes sense. [6243ff2](https://github.com/gravitational/teleport/pull/49789/commits/6243ff2009e9143c33e099934d409e1ea7f881ae) should allow the trusted_cluster labels to be modified now.

**Review by espadolini — APPROVED** [review]

**Comment by vintage-maeve** [comment]
Thanks for putting this feature together!
Based on this PR, if I want to update only the `attribute_to_roles` section of the TC, would that require rejoining or having an active invitation token?

**Review by rosstimothy — APPROVED** [review]

**Comment by bernardjkim** [comment]
I've updated the PR to also include `CreateTrustedClusterV2` and `UpdateTrustedClusterV2` in response to https://github.com/gravitational/teleport/pull/49920#discussion_r1875781415

**Review by rosstimothy — CHANGES_REQUESTED** [review]
Instead of adding V2 RPCs to the auth service can we please add new RPCs to the trust service?

**Comment by bernardjkim** [comment]
I've moved the V2 RPCs to the trust service. I've left the logic in auth service for now to limit the size and scope of this PR.

File: lib/auth/trustedcluster.go [hunk 14]
```diff
@@ -67,29 +119,27 @@ func (a *Server) UpsertTrustedCluster(ctx context.Context, tc types.TrustedClust
...
 	if existingCluster == nil {
-		return a.createTrustedCluster(ctx, tc)
+		return a.createTrustedCluster(ctx, tc, validateName)
 	}
 
-	if err := existingCluster.CanChangeStateTo(tc); err != nil {
-		return nil, trace.Wrap(err)
-	}
+	updated, err := a.updateTrustedCluster(ctx, tc, existingCluster)
+	return updated, trace.Wrap(err)
```
Comment (espadolini) [inline lib/auth/trustedcluster.go:126]: Should we give this a handful of tries? UpsertFoo should basically never fail due to concurrent writes, but this logic is liable to fail on a race between an upsert and a delete, or two upserts on a new trusted cluster.

File: lib/auth/trustedcluster.go [hunk 15]
```diff
@@ -46,6 +47,57 @@ import (
...
+	}
+
+	_, err := a.GetTrustedCluster(ctx, tc.GetName())
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+
+	if err == nil {
+		return nil, trace.AlreadyExists("trusted cluster with name %s already exists", tc.GetName())
+	}
```
[the remark this answers is not in the document]
Comment (bernardjkim) [reply lib/auth/trustedcluster.go]: Yeah, we can remove this check though. I don't see any problems with deferring until the backend write to check if cluster already exists.

File: lib/auth/trustedcluster.go (hunk 14 above)
Comment (bernardjkim) [reply lib/auth/trustedcluster.go]: That is a good point. I think we can defer this to a separate PR though. We'd also be modifying the behavior of the existing Upsert method.

**Review by rosstimothy — APPROVED** [review]

**Comment by bernardjkim** [comment]
Thanks for all the good feedback!
